R/model.R

```diff
@@ -369,22 +374,152 @@ compile_method <- function(quiet = TRUE,
     include_paths <- private$precompile_include_paths_
   }
-  if (is.null(dir) && !is.null(private$dir_)) {
-    dir <- absolute_path(private$dir_)
-  } else if (!is.null(dir)) {
-    dir <- absolute_path(dir)
-  }
-  if (!is.null(dir)) {
-    dir <- repair_path(dir)
-    checkmate::assert_directory_exists(dir, access = "rw")
-  }
 
   # temporary deprecation warnings
   if (isTRUE(threads)) {
     warning("'threads' is deprecated. Please use 'cpp_options = list(stan_threads = TRUE)' instead.")
     cpp_options[["stan_threads"]] <- TRUE
   }
 
+  model_name <- sub(" ", "_", paste0(strip_ext(basename(self$stan_file())), "_model"))
+
+  stancflags_val <- NULL
+  if (!is.null(include_paths)) {
+    checkmate::assert_directory_exists(include_paths, access = "r")
+    include_paths <- absolute_path(include_paths)
+    include_paths <- paste0(include_paths, collapse = ",")
+    if (cmdstan_version() >= "2.24") {
+      include_paths_flag <- " --include-paths="
+    } else {
+      include_paths_flag <- " --include_paths="
+    }
+    stancflags_val <- paste0(stancflags_val, include_paths_flag, include_paths, " ")
+  }
+
+  if (!is.null(cpp_options$stan_opencl)) {
+    stanc_options[["use-opencl"]] <- TRUE
+  }
+  if (is.null(stanc_options[["name"]])) {
+    stanc_options[["name"]] <- model_name
+  }
+
+  stanc_built_options <- c()
+  for (i in seq_len(length(stanc_options))) {
+    option_name <- names(stanc_options)[i]
+    if (isTRUE(as.logical(stanc_options[[i]]))) {
+      stanc_built_options <- c(stanc_built_options, paste0("--", option_name))
+    } else {
+      stanc_options[[i]] <- gsub(' ','\ ',stanc_options[[i]],fixed=TRUE)
+      stanc_built_options <- c(stanc_built_options, paste0("--", option_name, "=", stanc_options[[i]]))
+    }
+  }
+
+  if (cmdstan_version() < "2.24") {
+    prepare_precompiled(cpp_options, quiet)
+  }
+
+  if (make_or_stanc == "make"){
+    stancflags_val <- paste0("STANCFLAGS += ", stancflags_val, paste0(stanc_built_options, collapse = " "))
+    these_args <- c(
+      cpp_options_to_compile_flags(cpp_options),
+      stancflags_val
+    )
+  } else {
+    stancflags_val <- trimws(c(stancflags_val, c(stanc_built_options)))
+    these_args <- c(stancflags_val)
+  }
+  return(these_args)
+}
+CmdStanModel$set("private", name = "generate_arg_strings", value = generate_arg_strings_method)
+
+run_autoformatter_method <- function(quiet = FALSE,
+                                include_paths = NULL,
+                                stanc_options = list()) {
+
+
+  temp_hpp_file <- tempfile(pattern = "model-", fileext = ".hpp")
+  stanc_options[["o"]] <- temp_hpp_file
+
+  arg_strings <-
+    private$generate_arg_strings(
+      quiet = quiet,
+      make_or_stanc = "stanc",
+      include_paths = include_paths,
+      stanc_options = stanc_options
+    )
+
+  # if pedantic-mode warnings have been requested, first run without --auto-format & with echo=!quiet
+  if ("--warn-pedantic" %in% arg_strings) {
```

Comment:
I think it might be better to make a utility function (not part of the CmdStanModel) to run stanc3 with arbitrary arguments on a supplied .stan file. 

You can use that in check_syntax and autoformatter, etc.

There are other flags that won't work with the auto-format, like the standalone functions etc.